ContributesGraphExtension with Anvil interop fails on ForScope annotation

### Description

When Anvil interop is enabled and a contributed binding with the same scope as a `@ContributesGraphExtension` graph extension has a `ForScope` qualifier, Metro fails with
```
e: org.jetbrains.kotlin.util.FileAnalysisException: Somewhere in file /Users/gabrielittner/opensource/khonshu/sample/simple/app/simple/src/main/kotlin/com/freeletics/khonshu/sample/app/AppComponent.kt: org.jetbrains.kotlin.utils.exceptions.KotlinIllegalArgumentExceptionWithAttachments: Expected expression 'FirPropertyAccessExpressionImpl' to be resolved
        at org.jetbrains.kotlin.util.AnalysisExceptionsKt.wrapIntoFileAnalysisExceptionIfNeeded(AnalysisExceptions.kt:62)
        at org.jetbrains.kotlin.fir.FirCliExceptionHandler.handleExceptionOnFileAnalysis(Utils.kt:251)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverTransformer.transformFile(FirSupertypesResolution.kt:892)
        at org.jetbrains.kotlin.fir.declarations.FirFile.transform(FirFile.kt:46)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirTransformerBasedResolveProcessor.processFile(FirResolveProcessor.kt:48)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirTotalResolveProcessor.process(FirTotalResolveProcessor.kt:36)
        at org.jetbrains.kotlin.fir.pipeline.AnalyseKt.runResolution(analyse.kt:24)
        at org.jetbrains.kotlin.fir.pipeline.FirUtilsKt.resolveAndCheckFir(firUtils.kt:76)
        at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFrontendPipelinePhase.executePhase(JvmFrontendPipelinePhase.kt:167)
        at org.jetbrains.kotlin.cli.pipeline.jvm.JvmFrontendPipelinePhase.executePhase(JvmFrontendPipelinePhase.kt:56)
        at org.jetbrains.kotlin.cli.pipeline.PipelinePhase.phaseBody(PipelinePhase.kt:68)
        at org.jetbrains.kotlin.cli.pipeline.PipelinePhase.phaseBody(PipelinePhase.kt:58)
        at org.jetbrains.kotlin.config.phaser.SimpleNamedCompilerPhase.phaseBody(CompilerPhase.kt:215)
        at org.jetbrains.kotlin.config.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:111)
        at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:28)
        at org.jetbrains.kotlin.config.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:62)
        at org.jetbrains.kotlin.cli.pipeline.AbstractCliPipeline.runPhasedPipeline(AbstractCliPipeline.kt:106)
        at org.jetbrains.kotlin.cli.pipeline.AbstractCliPipeline.execute(AbstractCliPipeline.kt:65)
        at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecutePhased(K2JVMCompiler.kt:61)
        at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecutePhased(K2JVMCompiler.kt:36)
        at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:80)
        at org.jetbrains.kotlin.cli.common.CLICompiler.exec(CLICompiler.kt:337)
        at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:466)
        at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:75)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.doCompile(IncrementalCompilerRunner.kt:514)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:431)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.tryCompileIncrementally$lambda$10$compile(IncrementalCompilerRunner.kt:258)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.tryCompileIncrementally(IncrementalCompilerRunner.kt:276)
        at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:128)
        at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execIncrementalCompiler(CompileServiceImpl.kt:678)
        at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execIncrementalCompiler(CompileServiceImpl.kt:92)
        at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1805)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:360)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:714)
        at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:598)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:844)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:721)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:720)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: org.jetbrains.kotlin.utils.exceptions.KotlinIllegalArgumentExceptionWithAttachments: Expected expression 'FirPropertyAccessExpressionImpl' to be resolved
        at org.jetbrains.kotlin.fir.types.FirTypeUtilsKt.getResolvedType(FirTypeUtils.kt:214)
        at dev.zacsweers.metro.compiler.fir.FirKt.computeAnnotationHash(fir.kt:360)
        at dev.zacsweers.metro.compiler.fir.MetroFirAnnotation.cachedHashKey_delegate$lambda$0(MetroFirAnnotation.kt:20)
        at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:98)
        at dev.zacsweers.metro.compiler.fir.MetroFirAnnotation.getCachedHashKey(MetroFirAnnotation.kt:20)
        at dev.zacsweers.metro.compiler.fir.MetroFirAnnotation.hashCode(MetroFirAnnotation.kt:38)
        at dev.zacsweers.metro.compiler.fir.FirTypeKey.hashCode(FirTypeKey.kt)
        at java.base/java.util.HashMap.hash(HashMap.java:338)
        at java.base/java.util.HashMap.put(HashMap.java:619)
        at dev.zacsweers.metro.compiler.fir.generators.ContributedInterfaceSupertypeGenerator.processRankBasedReplacements(ContributedInterfaceSupertypeGenerator.kt:532)
        at dev.zacsweers.metro.compiler.fir.generators.ContributedInterfaceSupertypeGenerator.computeAdditionalSupertypes(ContributedInterfaceSupertypeGenerator.kt:297)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.addSupertypesFromExtensions(FirSupertypesResolution.kt:495)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes$lambda$15(FirSupertypesResolution.kt:472)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:402)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:445)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypes(FirSupertypesResolution.kt:325)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypes$default(FirSupertypesResolution.kt:318)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypesForOuterClass(FirSupertypesResolution.kt:315)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.calculateScopes(FirSupertypesResolution.kt:299)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopeForNestedClasses$lambda$2(FirSupertypesResolution.kt:279)
        at org.jetbrains.kotlin.fir.resolve.transformers.SupertypeComputationSession.getOrPutScopeForStaticNestedClasses(FirSupertypesResolution.kt:635)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopeForNestedClasses(FirSupertypesResolution.kt:278)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopes(FirSupertypesResolution.kt:359)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:387)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:445)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypes(FirSupertypesResolution.kt:325)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypes$default(FirSupertypesResolution.kt:318)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypesForOuterClass(FirSupertypesResolution.kt:315)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.calculateScopes(FirSupertypesResolution.kt:299)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopeForNestedClasses$lambda$2(FirSupertypesResolution.kt:279)
        at org.jetbrains.kotlin.fir.resolve.transformers.SupertypeComputationSession.getOrPutScopeForStaticNestedClasses(FirSupertypesResolution.kt:635)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopeForNestedClasses(FirSupertypesResolution.kt:278)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopes(FirSupertypesResolution.kt:359)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:387)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:445)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypes(FirSupertypesResolution.kt:325)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypes(FirSupertypesResolution.kt:331)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypes$default(FirSupertypesResolution.kt:318)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveAllSupertypesForOuterClass(FirSupertypesResolution.kt:315)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.calculateScopes(FirSupertypesResolution.kt:299)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopeForNestedClasses$lambda$2(FirSupertypesResolution.kt:279)
        at org.jetbrains.kotlin.fir.resolve.transformers.SupertypeComputationSession.getOrPutScopeForStaticNestedClasses(FirSupertypesResolution.kt:635)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopeForNestedClasses(FirSupertypesResolution.kt:278)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.prepareScopes(FirSupertypesResolution.kt:359)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:387)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.resolveSpecificClassLikeSupertypes(FirSupertypesResolution.kt:445)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitRegularClass(FirSupertypesResolution.kt:425)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitRegularClass(FirSupertypesResolution.kt:226)
        at org.jetbrains.kotlin.fir.declarations.FirRegularClass.accept(FirRegularClass.kt:48)
        at org.jetbrains.kotlin.fir.declarations.impl.FirRegularClassImpl.acceptChildren(FirRegularClassImpl.kt:63)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitDeclarationContent(FirSupertypesResolution.kt:413)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitRegularClass(FirSupertypesResolution.kt:426)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitRegularClass(FirSupertypesResolution.kt:226)
        at org.jetbrains.kotlin.fir.declarations.FirRegularClass.accept(FirRegularClass.kt:48)
        at org.jetbrains.kotlin.fir.declarations.impl.FirFileImpl.acceptChildren(FirFileImpl.kt:57)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitDeclarationContent(FirSupertypesResolution.kt:413)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitFile(FirSupertypesResolution.kt:593)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverVisitor.visitFile(FirSupertypesResolution.kt:226)
        at org.jetbrains.kotlin.fir.declarations.FirFile.accept(FirFile.kt:42)
        at org.jetbrains.kotlin.fir.resolve.transformers.FirSupertypeResolverTransformer.transformFile(FirSupertypesResolution.kt:71)
        ... 44 more
```

### Self-contained Reproducer

Note: the reproducer uses Metro annotations but I ran into this originally with the Anvil annotations for binding/qualifier/scope.
```kotlin
import dev.zacsweers.metro.ContributesBinding
import dev.zacsweers.metro.ContributesGraphExtension
import dev.zacsweers.metro.DependencyGraph
import dev.zacsweers.metro.ForScope
import dev.zacsweers.metro.Inject
import dev.zacsweers.metro.SingleIn
import dev.zacsweers.metro.binding

class Parent
class Child

@SingleIn(Parent::class)
@DependencyGraph(scope = Parent::class, isExtendable = true)
interface ParentComponent {
    @DependencyGraph.Factory
    interface Factory {
        fun create(): ParentComponent
    }
}

@ContributesGraphExtension(Child::class)
interface ChildComponent {
    @ContributesGraphExtension.Factory(Parent::class)
    interface Factory {
        fun create(): ChildComponent
    }
}

interface ActivityNavigator

@Inject
@ForScope(Child::class)
@SingleIn(Child::class)
@ContributesBinding(Child::class, binding<ActivityNavigator>())
class BottomSheetNavigator : ActivityNavigator
```

### Metro version

0.3.0-SNAPSHOT

### Context

_No response_

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/MetroAnnotations.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/MetroAnnotations.kt
@@ -370,7 +370,8 @@ private fun FirBasedSymbol<*>.metroAnnotations(
             continue
           }
           in ids.assistedAnnotations -> {
-            assisted = expectNullAndSet("assisted", assisted, MetroFirAnnotation(annotation))
+            assisted =
+              expectNullAndSet("assisted", assisted, MetroFirAnnotation(annotation, session))
             continue
           }
         }
@@ -402,7 +403,8 @@ private fun FirBasedSymbol<*>.metroAnnotations(
             continue
           }
           in ids.multibindsAnnotations -> {
-            multibinds = expectNullAndSet("multibinds", assisted, MetroFirAnnotation(annotation))
+            multibinds =
+              expectNullAndSet("multibinds", assisted, MetroFirAnnotation(annotation, session))
             continue
           }
           Symbols.ClassIds.Composable -> {
@@ -435,13 +437,13 @@ private fun FirBasedSymbol<*>.metroAnnotations(
     }
 
     if (annotationClass.isAnnotatedWithAny(session, ids.scopeAnnotations)) {
-      scope = expectNullAndSet("scope", scope, MetroFirAnnotation(annotation))
+      scope = expectNullAndSet("scope", scope, MetroFirAnnotation(annotation, session))
       continue
     } else if (annotationClass.isAnnotatedWithAny(session, ids.qualifierAnnotations)) {
-      qualifier = expectNullAndSet("qualifier", qualifier, MetroFirAnnotation(annotation))
+      qualifier = expectNullAndSet("qualifier", qualifier, MetroFirAnnotation(annotation, session))
       continue
     } else if (annotationClass.isAnnotatedWithAny(session, ids.mapKeyAnnotations)) {
-      mapKeys += MetroFirAnnotation(annotation)
+      mapKeys += MetroFirAnnotation(annotation, session)
       continue
     }
   }

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/MetroFirAnnotation.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/MetroFirAnnotation.kt
@@ -5,19 +5,25 @@ package dev.zacsweers.metro.compiler.fir
 import dev.zacsweers.metro.compiler.appendIterableWith
 import dev.zacsweers.metro.compiler.md5base64
 import dev.zacsweers.metro.compiler.unsafeLazy
+import org.jetbrains.kotlin.fir.FirSession
 import org.jetbrains.kotlin.fir.expressions.FirAnnotationCall
 import org.jetbrains.kotlin.fir.expressions.FirExpression
 import org.jetbrains.kotlin.fir.expressions.FirGetClassCall
 import org.jetbrains.kotlin.fir.expressions.FirLiteralExpression
 import org.jetbrains.kotlin.fir.expressions.FirResolvedQualifier
 import org.jetbrains.kotlin.fir.expressions.arguments
+import org.jetbrains.kotlin.fir.extensions.FirSupertypeGenerationExtension.TypeResolveService
 import org.jetbrains.kotlin.fir.types.renderReadable
 import org.jetbrains.kotlin.fir.types.renderReadableWithFqNames
 import org.jetbrains.kotlin.fir.types.resolvedType
 import org.jetbrains.kotlin.types.ConstantValueKind
 
-internal class MetroFirAnnotation(val fir: FirAnnotationCall) {
-  private val cachedHashKey by unsafeLazy { fir.computeAnnotationHash() }
+internal class MetroFirAnnotation(
+  val fir: FirAnnotationCall,
+  session: FirSession,
+  typeResolver: TypeResolveService? = null,
+) {
+  private val cachedHashKey by unsafeLazy { fir.computeAnnotationHash(session, typeResolver) }
   private val cachedToString by unsafeLazy {
     buildString { renderAsAnnotation(fir, simple = false) }
   }

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/fir.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/fir.kt
@@ -355,24 +355,30 @@ internal inline fun FirClass.singleAbstractFunction(
  * Computes a hash key for this annotation instance composed of its underlying type and value
  * arguments.
  */
-internal fun FirAnnotationCall.computeAnnotationHash(): Int {
+internal fun FirAnnotationCall.computeAnnotationHash(
+  session: FirSession,
+  typeResolver: TypeResolveService? = null,
+): Int {
   return Objects.hash(
-    resolvedType.classId,
+    toAnnotationClassIdSafe(session),
     arguments
-      .map {
-        when (it) {
-          is FirLiteralExpression -> it.value
+      .map { arg ->
+        when (arg) {
+          is FirLiteralExpression -> arg.value
           is FirGetClassCall -> {
-            val argument = it.argument
-            if (argument is FirResolvedQualifier) {
-              argument.classId
-            } else {
-              argument.resolvedType.classId
-            }
+            typeResolver?.let { arg.resolvedClassArgumentTarget(it)?.classId }
+              ?: run {
+                val argument = arg.argument
+                if (argument is FirResolvedQualifier) {
+                  argument.classId
+                } else {
+                  argument.resolvedType.classId
+                }
+              }
           }
           // Enum entry reference
           is FirPropertyAccessExpression -> {
-            it.calleeReference
+            arg.calleeReference
               .toResolvedPropertySymbol()
               ?.receiverParameter
               ?.typeRef
@@ -380,7 +386,7 @@ internal fun FirAnnotationCall.computeAnnotationHash(): Int {
               ?.classId
           }
           else -> {
-            error("Unexpected annotation argument type: ${it::class.java} - ${it.render()}")
+            error("Unexpected annotation argument type: ${arg::class.java} - ${arg.render()}")
           }
         }
       }
@@ -609,8 +615,12 @@ internal inline fun FirConstructorSymbol.validateVisibility(
 internal fun FirBasedSymbol<*>.qualifierAnnotation(session: FirSession): MetroFirAnnotation? =
   annotations.qualifierAnnotation(session)
 
-internal fun List<FirAnnotation>.qualifierAnnotation(session: FirSession): MetroFirAnnotation? =
-  asSequence().annotationAnnotatedWithAny(session, session.classIds.qualifierAnnotations)
+internal fun List<FirAnnotation>.qualifierAnnotation(
+  session: FirSession,
+  typeResolver: TypeResolveService? = null,
+): MetroFirAnnotation? =
+  asSequence()
+    .annotationAnnotatedWithAny(session, session.classIds.qualifierAnnotations, typeResolver)
 
 internal fun FirBasedSymbol<*>.mapKeyAnnotation(session: FirSession): MetroFirAnnotation? =
   annotations.mapKeyAnnotation(session)
@@ -631,18 +641,20 @@ internal fun Sequence<FirAnnotation>.scopeAnnotations(
 internal fun Sequence<FirAnnotation>.annotationAnnotatedWithAny(
   session: FirSession,
   names: Set<ClassId>,
+  typeResolver: TypeResolveService? = null,
 ): MetroFirAnnotation? {
-  return annotationsAnnotatedWithAny(session, names).firstOrNull()
+  return annotationsAnnotatedWithAny(session, names, typeResolver).firstOrNull()
 }
 
 internal fun Sequence<FirAnnotation>.annotationsAnnotatedWithAny(
   session: FirSession,
   names: Set<ClassId>,
+  typeResolver: TypeResolveService? = null,
 ): Sequence<MetroFirAnnotation> {
   return filter { it.isResolved }
     .filterIsInstance<FirAnnotationCall>()
     .filter { annotationCall -> annotationCall.isAnnotatedWithAny(session, names) }
-    .map { MetroFirAnnotation(it) }
+    .map { MetroFirAnnotation(it, session, typeResolver) }
 }
 
 internal fun FirAnnotationCall.isQualifier(session: FirSession): Boolean {

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributedInterfaceSupertypeGenerator.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributedInterfaceSupertypeGenerator.kt
@@ -361,7 +361,10 @@ internal class ContributedInterfaceSupertypeGenerator(session: FirSession) :
 
                 ContributedBinding(
                   contributingType,
-                  FirTypeKey(boundType, contributingType.annotations.qualifierAnnotation(session)),
+                  FirTypeKey(
+                    boundType,
+                    contributingType.annotations.qualifierAnnotation(session, typeResolver),
+                  ),
                   annotation.rankValue(),
                 )
               }

--- a/compiler/src/test/kotlin/dev/zacsweers/metro/compiler/fir/AnvilInteropTest.kt
+++ b/compiler/src/test/kotlin/dev/zacsweers/metro/compiler/fir/AnvilInteropTest.kt
@@ -30,7 +30,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
             .trimIndent()
         ),
-        options = metroOptions.withAnvilContributesBinding(),
+        options = metroOptions.withAnvilInterop(),
       )
 
     compile(
@@ -47,7 +47,7 @@ class AnvilInteropTest : MetroCompilerTest() {
           .trimIndent()
       ),
       previousCompilationResult = previousCompilation,
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -69,7 +69,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
             .trimIndent()
         ),
-        options = metroOptions.withAnvilContributesBinding(),
+        options = metroOptions.withAnvilInterop(),
       )
 
     compile(
@@ -86,7 +86,7 @@ class AnvilInteropTest : MetroCompilerTest() {
           .trimIndent()
       ),
       previousCompilationResult = libCompilation,
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -118,7 +118,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -155,7 +155,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -203,7 +203,7 @@ class AnvilInteropTest : MetroCompilerTest() {
           .trimIndent()
       ),
       previousCompilationResult = libCompilation,
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -232,7 +232,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -261,7 +261,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -295,7 +295,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -326,7 +326,7 @@ class AnvilInteropTest : MetroCompilerTest() {
           """
             .trimIndent()
         ),
-        options = metroOptions.withAnvilContributesBinding(),
+        options = metroOptions.withAnvilInterop(),
       )
 
     compile(
@@ -354,7 +354,7 @@ class AnvilInteropTest : MetroCompilerTest() {
       previousCompilationResult = previousCompilation,
       options =
         metroOptions
-          .withAnvilContributesBinding()
+          .withAnvilInterop()
           .copy(customQualifierAnnotations = setOf(ClassId.fromString("test/ThirdPartyQualifier"))),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
@@ -388,7 +388,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
     ) {
       val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
       val contributedInterface = graph.callProperty<Any>("contributedInterface")
@@ -420,7 +420,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
       expectedExitCode = ExitCode.COMPILATION_ERROR,
     ) {
       assertDiagnostics(
@@ -493,7 +493,7 @@ class AnvilInteropTest : MetroCompilerTest() {
         """
           .trimIndent()
       ),
-      options = metroOptions.withAnvilContributesBinding(),
+      options = metroOptions.withAnvilInterop(),
       expectedExitCode = ExitCode.INTERNAL_ERROR,
     ) {
       assertThat(messages)
@@ -506,10 +506,54 @@ class AnvilInteropTest : MetroCompilerTest() {
     }
   }
 
-  private fun MetroOptions.withAnvilContributesBinding(): MetroOptions {
-    return metroOptions.copy(
+  // https://github.com/ZacSweers/metro/issues/388
+  @Test
+  fun `ContributesSubcomponent test`() {
+    compile(
+      source(
+        """
+          class Parent
+          class Child
+
+          @SingleIn(Parent::class)
+          @DependencyGraph(scope = Parent::class, isExtendable = true)
+          interface ParentComponent {
+              @DependencyGraph.Factory
+              interface Factory {
+                  fun create(): ParentComponent
+              }
+          }
+
+          @com.squareup.anvil.annotations.ContributesSubcomponent(Child::class, parentScope = Parent::class)
+          interface ChildComponent {
+              @ContributesGraphExtension.Factory(Parent::class)
+              interface Factory {
+                  fun create(): ChildComponent
+              }
+          }
+
+          interface ActivityNavigator
+
+          @Inject
+          @ForScope(Child::class)
+          @SingleIn(Child::class)
+          @com.squareup.anvil.annotations.ContributesBinding(Child::class, boundType = ActivityNavigator::class)
+          class BottomSheetNavigator : ActivityNavigator
+        """
+          .trimIndent()
+      ),
+      options = metroOptions.withAnvilInterop(),
+    ) {
+      println("asdf")
+    }
+  }
+
+  private fun MetroOptions.withAnvilInterop(): MetroOptions {
+    return copy(
       customContributesBindingAnnotations =
         setOf(ClassId.fromString("com/squareup/anvil/annotations/ContributesBinding")),
+      customContributesGraphExtensionAnnotations =
+        setOf(ClassId.fromString("com/squareup/anvil/annotations/ContributesSubcomponent")),
       enableDaggerAnvilInterop = true,
     )
   }

